core/src/services/gdrive/core.rs

```diff
@@ -0,0 +1,202 @@
+                query
+            ));
+            req = req.header(header::AUTHORIZATION, &auth_header_content);
+            let req = req
+                .body(AsyncBody::default())
+                .map_err(new_request_build_error)
+                .unwrap();
+
+            let resp = self.client.send(req).await.unwrap();
+
+            let body_value: GdriveFileList =
+                serde_json::from_slice(&resp.into_body().bytes().await.unwrap()).unwrap();
+            parent_id = String::from(body_value.files[0].id.as_str());
+        }
+
+        let mut cache_guard = self.path_2_id.lock().await;
+        cache_guard.insert(path, parent_id.clone());
+
+        parent_id
+    }
+
+    pub async fn gdrive_get(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let url: String = format!(
+            "https://www.googleapis.com/drive/v3/files/{}?alt=media",
+            self.get_file_id_by_path(path).await
+        );
+
+        let auth_header_content = format!("Bearer {}", self.access_token);
+        let mut req = Request::get(&url);
+        req = req.header(header::AUTHORIZATION, auth_header_content);
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn gdrive_update(
+        &self,
+        path: &str,
+        size: Option<usize>,
+        content_type: Option<&str>,
+        body: AsyncBody,
+    ) -> Result<Response<IncomingAsyncBody>> {
+        let url = format!(
+            "https://www.googleapis.com/upload/drive/v3/files/{}",
+            self.get_file_id_by_path(path).await
+        );
+
+        let mut req = Request::patch(&url);
+
+        let auth_header_content = format!("Bearer {}", self.access_token);
+        req = req.header(header::AUTHORIZATION, auth_header_content);
+
+        if let Some(size) = size {
+            req = req.header(header::CONTENT_LENGTH, size)
+        }
+
+        if let Some(mime) = content_type {
+            req = req.header(header::CONTENT_TYPE, mime)
+        }
+
+        let req = req.body(body).map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn gdrive_delete(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let url = format!(
+            "https://www.googleapis.com/drive/v3/files/{}",
+            self.get_file_id_by_path(path).await
+        );
+
+        let mut req = Request::delete(&url);
+
+        let auth_header_content = format!("Bearer {}", self.access_token);
+        req = req.header(header::AUTHORIZATION, auth_header_content);
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+}
+
+#[derive(Deserialize)]
+struct GdriveFile {
+    id: String,
+}
+
+#[derive(Deserialize)]
+struct GdriveFileList {
```

Comment:
ditto, I guess we should link to https://developers.google.com/drive/api/reference/rest/v3/files/list?